// angular2-token/esm2015/a2t-ui/a2t-reset-password/a2t-reset-password.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component } from '@angular/core';
import { Angular2TokenService } from '../../angular2-token.service';
import { A2tFormService } from '../a2t-shared/a2t-form/a2t-form.service';
import { RESET_PASSWORD_FORM } from '../a2t-ui.forms';
export class A2tResetPasswordComponent {
    /**
     * @param {?} _formService
     * @param {?} _sessionService
     */
    constructor(_formService, _sessionService) {
        this._formService = _formService;
        this._sessionService = _sessionService;
        this._emailSend = false;
        this._formService.initForm(RESET_PASSWORD_FORM);
        this._formService.submit$.subscribe((data) => this._sessionService.resetPassword(data).subscribe(res => this._handleSuccess(), error => this._handleError()));
    }
    /**
     * @return {?}
     */
    _handleSuccess() {
        this._emailSend = true;
    }
    /**
     * @return {?}
     */
    _handleError() {
        this._emailSend = true;
    }
}
A2tResetPasswordComponent.decorators = [
    { type: Component, args: [{
                selector: 'a2t-reset-password',
                providers: [A2tFormService],
                template: `
        <a2t-headline *ngIf="!_emailSend">Reset your Password</a2t-headline>
        <a2t-form *ngIf="!_emailSend">Reset Password</a2t-form>
        <p class="email-send-text" *ngIf="_emailSend">
            If the entered email is registered we will send instruction on how to reset your password.
        </p>
        <a2t-links *ngIf="!_emailSend" case="reset-password"></a2t-links>
    `,
                styles: [`
        .email-send-text {
            background-color: #72c380;
            color: white;
            font-size: 16pt;
            text-align: center;
            padding: 20px;
            border-radius: 3px;
        }
    `]
            },] },
];
/** @nocollapse */
A2tResetPasswordComponent.ctorParameters = () => [
    { type: A2tFormService },
    { type: Angular2TokenService }
];
function A2tResetPasswordComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    A2tResetPasswordComponent.prototype._emailSend;
    /** @type {?} */
    A2tResetPasswordComponent.prototype._formService;
    /** @type {?} */
    A2tResetPasswordComponent.prototype._sessionService;
}
