// angular2-token/esm2015/a2t-ui/a2t-shared/a2t-error/a2t-error.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input } from '@angular/core';
export class A2tErrorComponent {
    constructor() { }
}
A2tErrorComponent.decorators = [
    { type: Component, args: [{
                selector: 'a2t-error',
                template: '<div *ngFor="let error of errors"><p>{{error}}</p></div>',
                styles: [`
        div {
            width: 100%;
            background-color: #df6564;
            color: white;
            font-weight: 300;
            font-size: 15px;
            padding: 10px 20px;
            border-radius: 3px;
            margin-bottom: 15px;
        }

        div > p {
            margin-bottom: 0;
        }
    `]
            },] },
];
/** @nocollapse */
A2tErrorComponent.ctorParameters = () => [];
A2tErrorComponent.propDecorators = {
    errors: [{ type: Input }]
};
function A2tErrorComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    A2tErrorComponent.prototype.errors;
}
