// angular2-token/esm5/a2t-ui/a2t-shared/a2t-error/a2t-error.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input } from '@angular/core';
var A2tErrorComponent = /** @class */ (function () {
    function A2tErrorComponent() {
    }
    A2tErrorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'a2t-error',
                    template: '<div *ngFor="let error of errors"><p>{{error}}</p></div>',
                    styles: ["\n        div {\n            width: 100%;\n            background-color: #df6564;\n            color: white;\n            font-weight: 300;\n            font-size: 15px;\n            padding: 10px 20px;\n            border-radius: 3px;\n            margin-bottom: 15px;\n        }\n\n        div > p {\n            margin-bottom: 0;\n        }\n    "]
                },] },
    ];
    /** @nocollapse */
    A2tErrorComponent.ctorParameters = function () { return []; };
    A2tErrorComponent.propDecorators = {
        errors: [{ type: Input }]
    };
    return A2tErrorComponent;
}());
export { A2tErrorComponent };
function A2tErrorComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    A2tErrorComponent.prototype.errors;
}
