// angular2-token/esm5/angular2-token.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
// Function Data
/**
 * @record
 */
export function SignInData() { }
function SignInData_tsickle_Closure_declarations() {
    /** @type {?} */
    SignInData.prototype.email;
    /** @type {?} */
    SignInData.prototype.password;
    /** @type {?|undefined} */
    SignInData.prototype.userType;
}
/**
 * @record
 */
export function RegisterData() { }
function RegisterData_tsickle_Closure_declarations() {
    /** @type {?} */
    RegisterData.prototype.email;
    /** @type {?} */
    RegisterData.prototype.password;
    /** @type {?} */
    RegisterData.prototype.passwordConfirmation;
    /** @type {?|undefined} */
    RegisterData.prototype.name;
    /** @type {?|undefined} */
    RegisterData.prototype.userType;
}
/**
 * @record
 */
export function RegisterData() { }
function RegisterData_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    [key:string]: string;
    */
}
;
/**
 * @record
 */
export function UpdatePasswordData() { }
function UpdatePasswordData_tsickle_Closure_declarations() {
    /** @type {?} */
    UpdatePasswordData.prototype.password;
    /** @type {?} */
    UpdatePasswordData.prototype.passwordConfirmation;
    /** @type {?|undefined} */
    UpdatePasswordData.prototype.passwordCurrent;
    /** @type {?|undefined} */
    UpdatePasswordData.prototype.userType;
    /** @type {?|undefined} */
    UpdatePasswordData.prototype.resetPasswordToken;
}
/**
 * @record
 */
export function ResetPasswordData() { }
function ResetPasswordData_tsickle_Closure_declarations() {
    /** @type {?} */
    ResetPasswordData.prototype.email;
    /** @type {?|undefined} */
    ResetPasswordData.prototype.userType;
}
/**
 * @record
 */
export function AuthData() { }
function AuthData_tsickle_Closure_declarations() {
    /** @type {?} */
    AuthData.prototype.accessToken;
    /** @type {?} */
    AuthData.prototype.client;
    /** @type {?} */
    AuthData.prototype.expiry;
    /** @type {?} */
    AuthData.prototype.tokenType;
    /** @type {?} */
    AuthData.prototype.uid;
}
/**
 * @record
 */
export function UserData() { }
function UserData_tsickle_Closure_declarations() {
    /** @type {?} */
    UserData.prototype.id;
    /** @type {?} */
    UserData.prototype.provider;
    /** @type {?} */
    UserData.prototype.uid;
    /** @type {?} */
    UserData.prototype.name;
    /** @type {?} */
    UserData.prototype.nickname;
    /** @type {?} */
    UserData.prototype.image;
    /** @type {?} */
    UserData.prototype.email;
}
/**
 * @record
 */
export function UserType() { }
function UserType_tsickle_Closure_declarations() {
    /** @type {?} */
    UserType.prototype.name;
    /** @type {?} */
    UserType.prototype.path;
}
/**
 * @record
 */
export function GlobalOptions() { }
function GlobalOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    GlobalOptions.prototype.headers;
}
/**
 * @record
 */
export function Angular2TokenOptions() { }
function Angular2TokenOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.apiBase;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.apiPath;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.signInPath;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.signInRedirect;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.signInStoredUrlStorageKey;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.signOutPath;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.validateTokenPath;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.signOutFailedValidate;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.deleteAccountPath;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.registerAccountPath;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.registerAccountCallback;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.updatePasswordPath;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.resetPasswordPath;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.resetPasswordCallback;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.userTypes;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.oAuthBase;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.oAuthPaths;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.oAuthCallbackPath;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.oAuthWindowType;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.oAuthWindowOptions;
    /** @type {?|undefined} */
    Angular2TokenOptions.prototype.globalOptions;
}
