// angular2-token/esm5/angular2-token.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable, Optional } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';
import { Http, Headers, Request, RequestMethod, RequestOptions } from '@angular/http';
import { interval, fromEvent } from 'rxjs';
import { share, pluck, filter } from 'rxjs/operators';
var Angular2TokenService = /** @class */ (function () {
    function Angular2TokenService(http, activatedRoute, router) {
        this.http = http;
        this.activatedRoute = activatedRoute;
        this.router = router;
    }
    Object.defineProperty(Angular2TokenService.prototype, "currentUserType", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.atCurrentUserType != null)
                return this.atCurrentUserType.name;
            else
                return null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Angular2TokenService.prototype, "currentUserData", {
        get: /**
         * @return {?}
         */
        function () {
            return this.atCurrentUserData;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Angular2TokenService.prototype, "currentAuthData", {
        get: /**
         * @return {?}
         */
        function () {
            return this.atCurrentAuthData;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Angular2TokenService.prototype, "currentAuthHeaders", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.atCurrentAuthData != null) {
                return new Headers({
                    'access-token': this.atCurrentAuthData.accessToken,
                    'client': this.atCurrentAuthData.client,
                    'expiry': this.atCurrentAuthData.expiry,
                    'token-type': this.atCurrentAuthData.tokenType,
                    'uid': this.atCurrentAuthData.uid
                });
            }
            return new Headers;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    Angular2TokenService.prototype.userSignedIn = /**
     * @return {?}
     */
    function () {
        return !!this.atCurrentAuthData;
    };
    /**
     * @return {?}
     */
    Angular2TokenService.prototype.canActivate = /**
     * @return {?}
     */
    function () {
        if (this.userSignedIn())
            return true;
        else {
            // Store current location in storage (usefull for redirection after signing in)
            if (this.atOptions.signInStoredUrlStorageKey) {
                localStorage.setItem(this.atOptions.signInStoredUrlStorageKey, window.location.pathname + window.location.search);
            }
            // Redirect user to sign in if signInRedirect is set
            if (this.router && this.atOptions.signInRedirect)
                this.router.navigate([this.atOptions.signInRedirect]);
            return false;
        }
    };
    // Inital configuration
    /**
     * @param {?=} options
     * @return {?}
     */
    Angular2TokenService.prototype.init = /**
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        var /** @type {?} */ defaultOptions = {
            apiPath: null,
            apiBase: null,
            signInPath: 'auth/sign_in',
            signInRedirect: null,
            signInStoredUrlStorageKey: null,
            signOutPath: 'auth/sign_out',
            validateTokenPath: 'auth/validate_token',
            signOutFailedValidate: false,
            registerAccountPath: 'auth',
            deleteAccountPath: 'auth',
            registerAccountCallback: window.location.href,
            updatePasswordPath: 'auth',
            resetPasswordPath: 'auth/password',
            resetPasswordCallback: window.location.href,
            userTypes: null,
            oAuthBase: window.location.origin,
            oAuthPaths: {
                github: 'auth/github'
            },
            oAuthCallbackPath: 'oauth_callback',
            oAuthWindowType: 'newWindow',
            oAuthWindowOptions: null,
            globalOptions: {
                headers: {
                    'Content-Type': 'application/json',
                    'Accept': 'application/json'
                }
            }
        };
        this.atOptions = (/** @type {?} */ (Object)).assign(defaultOptions, options);
        this.tryLoadAuthData();
    };
    /**
     *
     * Actions
     *
     */
    // Register request
    /**
     *
     * Actions
     *
     * @param {?} registerData
     * @return {?}
     */
    Angular2TokenService.prototype.registerAccount = /**
     *
     * Actions
     *
     * @param {?} registerData
     * @return {?}
     */
    function (registerData) {
        if (registerData.userType == null)
            this.atCurrentUserType = null;
        else {
            this.atCurrentUserType = this.getUserTypeByName(registerData.userType);
            delete registerData.userType;
        }
        if (registerData["password_confirmation"] == null &&
            registerData.passwordConfirmation != null) {
            registerData["password_confirmation"] = registerData.passwordConfirmation;
            delete registerData.passwordConfirmation;
        }
        registerData["confirm_success_url"] = this.atOptions.registerAccountCallback;
        return this.post(this.getUserPath() + this.atOptions.registerAccountPath, JSON.stringify(registerData));
    };
    // Delete Account
    /**
     * @return {?}
     */
    Angular2TokenService.prototype.deleteAccount = /**
     * @return {?}
     */
    function () {
        return this.delete(this.getUserPath() + this.atOptions.deleteAccountPath);
    };
    // Sign in request and set storage
    /**
     * @param {?} signInData
     * @return {?}
     */
    Angular2TokenService.prototype.signIn = /**
     * @param {?} signInData
     * @return {?}
     */
    function (signInData) {
        var _this = this;
        if (signInData.userType == null)
            this.atCurrentUserType = null;
        else
            this.atCurrentUserType = this.getUserTypeByName(signInData.userType);
        var /** @type {?} */ body = JSON.stringify({
            email: signInData.email,
            password: signInData.password
        });
        var /** @type {?} */ observ = this.post(this.getUserPath() + this.atOptions.signInPath, body);
        observ.subscribe(function (res) { return _this.atCurrentUserData = res.json().data; }, function (_error) { return null; });
        return observ;
    };
    /**
     * @param {?} oAuthType
     * @return {?}
     */
    Angular2TokenService.prototype.signInOAuth = /**
     * @param {?} oAuthType
     * @return {?}
     */
    function (oAuthType) {
        var /** @type {?} */ oAuthPath = this.getOAuthPath(oAuthType);
        var /** @type {?} */ callbackUrl = window.location.origin + "/" + this.atOptions.oAuthCallbackPath;
        var /** @type {?} */ oAuthWindowType = this.atOptions.oAuthWindowType;
        var /** @type {?} */ authUrl = this.getOAuthUrl(oAuthPath, callbackUrl, oAuthWindowType);
        if (oAuthWindowType == 'newWindow') {
            var /** @type {?} */ oAuthWindowOptions = this.atOptions.oAuthWindowOptions;
            var /** @type {?} */ windowOptions = '';
            if (oAuthWindowOptions) {
                for (var /** @type {?} */ key in oAuthWindowOptions) {
                    windowOptions += "," + key + "=" + oAuthWindowOptions[key];
                }
            }
            var /** @type {?} */ popup = window.open(authUrl, '_blank', "closebuttoncaption=Cancel" + windowOptions);
            return this.requestCredentialsViaPostMessage(popup);
        }
        else if (oAuthWindowType == 'sameWindow') {
            window.location.href = authUrl;
        }
        else {
            throw "Unsupported oAuthWindowType \"" + oAuthWindowType + "\"";
        }
    };
    /**
     * @return {?}
     */
    Angular2TokenService.prototype.processOAuthCallback = /**
     * @return {?}
     */
    function () {
        this.getAuthDataFromParams();
    };
    // Sign out request and delete storage
    /**
     * @return {?}
     */
    Angular2TokenService.prototype.signOut = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ observ = this.delete(this.getUserPath() + this.atOptions.signOutPath);
        localStorage.removeItem('accessToken');
        localStorage.removeItem('client');
        localStorage.removeItem('expiry');
        localStorage.removeItem('tokenType');
        localStorage.removeItem('uid');
        this.atCurrentAuthData = null;
        this.atCurrentUserType = null;
        this.atCurrentUserData = null;
        return observ;
    };
    // Validate token request
    /**
     * @return {?}
     */
    Angular2TokenService.prototype.validateToken = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var /** @type {?} */ observ = this.get(this.getUserPath() + this.atOptions.validateTokenPath);
        observ.subscribe(function (res) { return _this.atCurrentUserData = res.json().data; }, function (error) {
            if (error.status === 401 && _this.atOptions.signOutFailedValidate) {
                _this.signOut();
            }
        });
        return observ;
    };
    // Update password request
    /**
     * @param {?} updatePasswordData
     * @return {?}
     */
    Angular2TokenService.prototype.updatePassword = /**
     * @param {?} updatePasswordData
     * @return {?}
     */
    function (updatePasswordData) {
        if (updatePasswordData.userType != null)
            this.atCurrentUserType = this.getUserTypeByName(updatePasswordData.userType);
        var /** @type {?} */ args;
        if (updatePasswordData.passwordCurrent == null) {
            args = {
                password: updatePasswordData.password,
                password_confirmation: updatePasswordData.passwordConfirmation
            };
        }
        else {
            args = {
                current_password: updatePasswordData.passwordCurrent,
                password: updatePasswordData.password,
                password_confirmation: updatePasswordData.passwordConfirmation
            };
        }
        if (updatePasswordData.resetPasswordToken) {
            args.reset_password_token = updatePasswordData.resetPasswordToken;
        }
        var /** @type {?} */ body = JSON.stringify(args);
        return this.put(this.getUserPath() + this.atOptions.updatePasswordPath, body);
    };
    // Reset password request
    /**
     * @param {?} resetPasswordData
     * @return {?}
     */
    Angular2TokenService.prototype.resetPassword = /**
     * @param {?} resetPasswordData
     * @return {?}
     */
    function (resetPasswordData) {
        if (resetPasswordData.userType == null)
            this.atCurrentUserType = null;
        else
            this.atCurrentUserType = this.getUserTypeByName(resetPasswordData.userType);
        var /** @type {?} */ body = JSON.stringify({
            email: resetPasswordData.email,
            redirect_url: this.atOptions.resetPasswordCallback
        });
        return this.post(this.getUserPath() + this.atOptions.resetPasswordPath, body);
    };
    /**
     *
     * HTTP Wrappers
     *
     */
    /**
     *
     * HTTP Wrappers
     *
     * @param {?} url
     * @param {?=} options
     * @return {?}
     */
    Angular2TokenService.prototype.get = /**
     *
     * HTTP Wrappers
     *
     * @param {?} url
     * @param {?=} options
     * @return {?}
     */
    function (url, options) {
        return this.request(this.mergeRequestOptionsArgs({
            url: this.getApiPath() + url,
            method: RequestMethod.Get
        }, options));
    };
    /**
     * @param {?} url
     * @param {?} body
     * @param {?=} options
     * @return {?}
     */
    Angular2TokenService.prototype.post = /**
     * @param {?} url
     * @param {?} body
     * @param {?=} options
     * @return {?}
     */
    function (url, body, options) {
        return this.request(this.mergeRequestOptionsArgs({
            url: this.getApiPath() + url,
            method: RequestMethod.Post,
            body: body
        }, options));
    };
    /**
     * @param {?} url
     * @param {?} body
     * @param {?=} options
     * @return {?}
     */
    Angular2TokenService.prototype.put = /**
     * @param {?} url
     * @param {?} body
     * @param {?=} options
     * @return {?}
     */
    function (url, body, options) {
        return this.request(this.mergeRequestOptionsArgs({
            url: this.getApiPath() + url,
            method: RequestMethod.Put,
            body: body
        }, options));
    };
    /**
     * @param {?} url
     * @param {?=} options
     * @return {?}
     */
    Angular2TokenService.prototype.delete = /**
     * @param {?} url
     * @param {?=} options
     * @return {?}
     */
    function (url, options) {
        return this.request(this.mergeRequestOptionsArgs({
            url: this.getApiPath() + url,
            method: RequestMethod.Delete
        }, options));
    };
    /**
     * @param {?} url
     * @param {?} body
     * @param {?=} options
     * @return {?}
     */
    Angular2TokenService.prototype.patch = /**
     * @param {?} url
     * @param {?} body
     * @param {?=} options
     * @return {?}
     */
    function (url, body, options) {
        return this.request(this.mergeRequestOptionsArgs({
            url: this.getApiPath() + url,
            method: RequestMethod.Patch,
            body: body
        }, options));
    };
    /**
     * @param {?} path
     * @param {?=} options
     * @return {?}
     */
    Angular2TokenService.prototype.head = /**
     * @param {?} path
     * @param {?=} options
     * @return {?}
     */
    function (path, options) {
        return this.request({
            method: RequestMethod.Head,
            url: this.getApiPath() + path
        });
    };
    /**
     * @param {?} url
     * @param {?=} options
     * @return {?}
     */
    Angular2TokenService.prototype.options = /**
     * @param {?} url
     * @param {?=} options
     * @return {?}
     */
    function (url, options) {
        return this.request(this.mergeRequestOptionsArgs({
            url: this.getApiPath() + url,
            method: RequestMethod.Options
        }, options));
    };
    // Construct and send Http request
    /**
     * @param {?} options
     * @return {?}
     */
    Angular2TokenService.prototype.request = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        var /** @type {?} */ baseRequestOptions;
        var /** @type {?} */ baseHeaders = this.atOptions.globalOptions.headers;
        // Get auth data from local storage
        this.getAuthDataFromStorage();
        // Merge auth headers to request if set
        if (this.atCurrentAuthData != null) {
            (/** @type {?} */ (Object)).assign(baseHeaders, {
                'access-token': this.atCurrentAuthData.accessToken,
                'client': this.atCurrentAuthData.client,
                'expiry': this.atCurrentAuthData.expiry,
                'token-type': this.atCurrentAuthData.tokenType,
                'uid': this.atCurrentAuthData.uid
            });
        }
        baseRequestOptions = new RequestOptions({
            headers: new Headers(baseHeaders)
        });
        // Merge standard and custom RequestOptions
        baseRequestOptions = baseRequestOptions.merge(options);
        var /** @type {?} */ response = this.http.request(new Request(baseRequestOptions)).pipe(share());
        this.handleResponse(response);
        return response;
    };
    /**
     * @param {?} options
     * @param {?=} addOptions
     * @return {?}
     */
    Angular2TokenService.prototype.mergeRequestOptionsArgs = /**
     * @param {?} options
     * @param {?=} addOptions
     * @return {?}
     */
    function (options, addOptions) {
        var /** @type {?} */ returnOptions = options;
        if (options)
            (/** @type {?} */ (Object)).assign(returnOptions, addOptions);
        return returnOptions;
    };
    /**
     * @param {?} response
     * @return {?}
     */
    Angular2TokenService.prototype.handleResponse = /**
     * @param {?} response
     * @return {?}
     */
    function (response) {
        var _this = this;
        response.subscribe(function (res) {
            _this.getAuthHeadersFromResponse(/** @type {?} */ (res));
        }, function (error) {
            _this.getAuthHeadersFromResponse(/** @type {?} */ (error));
        });
    };
    /**
     *
     * Get Auth Data
     *
     * @return {?}
     */
    Angular2TokenService.prototype.tryLoadAuthData = /**
     *
     * Get Auth Data
     *
     * @return {?}
     */
    function () {
        var /** @type {?} */ userType = this.getUserTypeByName(localStorage.getItem('userType'));
        if (userType)
            this.atCurrentUserType = userType;
        this.getAuthDataFromStorage();
        if (this.activatedRoute)
            this.getAuthDataFromParams();
        if (this.atCurrentAuthData)
            this.validateToken();
    };
    /**
     * @param {?} data
     * @return {?}
     */
    Angular2TokenService.prototype.getAuthHeadersFromResponse = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var /** @type {?} */ headers = data.headers;
        var /** @type {?} */ authData = {
            accessToken: headers.get('access-token'),
            client: headers.get('client'),
            expiry: headers.get('expiry'),
            tokenType: headers.get('token-type'),
            uid: headers.get('uid')
        };
        this.setAuthData(authData);
    };
    /**
     * @param {?} data
     * @return {?}
     */
    Angular2TokenService.prototype.getAuthDataFromPostMessage = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var /** @type {?} */ authData = {
            accessToken: data['auth_token'],
            client: data['client_id'],
            expiry: data['expiry'],
            tokenType: 'Bearer',
            uid: data['uid']
        };
        this.setAuthData(authData);
    };
    /**
     * @return {?}
     */
    Angular2TokenService.prototype.getAuthDataFromStorage = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ authData = {
            accessToken: localStorage.getItem('accessToken'),
            client: localStorage.getItem('client'),
            expiry: localStorage.getItem('expiry'),
            tokenType: localStorage.getItem('tokenType'),
            uid: localStorage.getItem('uid')
        };
        if (this.checkAuthData(authData))
            this.atCurrentAuthData = authData;
    };
    /**
     * @return {?}
     */
    Angular2TokenService.prototype.getAuthDataFromParams = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.activatedRoute.queryParams)
            // Fix for Testing, needs to be removed later
            this.activatedRoute.queryParams.subscribe(function (queryParams) {
                var /** @type {?} */ authData = {
                    accessToken: queryParams['token'] || queryParams['auth_token'],
                    client: queryParams['client_id'],
                    expiry: queryParams['expiry'],
                    tokenType: 'Bearer',
                    uid: queryParams['uid']
                };
                if (_this.checkAuthData(authData))
                    _this.atCurrentAuthData = authData;
            });
    };
    /**
     *
     * Set Auth Data
     *
     * @param {?} authData
     * @return {?}
     */
    Angular2TokenService.prototype.setAuthData = /**
     *
     * Set Auth Data
     *
     * @param {?} authData
     * @return {?}
     */
    function (authData) {
        if (this.checkAuthData(authData)) {
            this.atCurrentAuthData = authData;
            localStorage.setItem('accessToken', authData.accessToken);
            localStorage.setItem('client', authData.client);
            localStorage.setItem('expiry', authData.expiry);
            localStorage.setItem('tokenType', authData.tokenType);
            localStorage.setItem('uid', authData.uid);
            if (this.atCurrentUserType != null)
                localStorage.setItem('userType', this.atCurrentUserType.name);
        }
    };
    /**
     *
     * Validate Auth Data
     *
     * @param {?} authData
     * @return {?}
     */
    Angular2TokenService.prototype.checkAuthData = /**
     *
     * Validate Auth Data
     *
     * @param {?} authData
     * @return {?}
     */
    function (authData) {
        if (authData.accessToken != null &&
            authData.client != null &&
            authData.expiry != null &&
            authData.tokenType != null &&
            authData.uid != null) {
            if (this.atCurrentAuthData != null)
                return authData.expiry >= this.atCurrentAuthData.expiry;
            else
                return true;
        }
        else {
            return false;
        }
    };
    /**
     *
     * Construct Paths / Urls
     *
     * @return {?}
     */
    Angular2TokenService.prototype.getUserPath = /**
     *
     * Construct Paths / Urls
     *
     * @return {?}
     */
    function () {
        if (this.atCurrentUserType == null)
            return '';
        else
            return this.atCurrentUserType.path + '/';
    };
    /**
     * @return {?}
     */
    Angular2TokenService.prototype.getApiPath = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ constructedPath = '';
        if (this.atOptions.apiBase != null)
            constructedPath += this.atOptions.apiBase + '/';
        if (this.atOptions.apiPath != null)
            constructedPath += this.atOptions.apiPath + '/';
        return constructedPath;
    };
    /**
     * @param {?} oAuthType
     * @return {?}
     */
    Angular2TokenService.prototype.getOAuthPath = /**
     * @param {?} oAuthType
     * @return {?}
     */
    function (oAuthType) {
        var /** @type {?} */ oAuthPath;
        oAuthPath = this.atOptions.oAuthPaths[oAuthType];
        if (oAuthPath == null)
            oAuthPath = "/auth/" + oAuthType;
        return oAuthPath;
    };
    /**
     * @param {?} oAuthPath
     * @param {?} callbackUrl
     * @param {?} windowType
     * @return {?}
     */
    Angular2TokenService.prototype.getOAuthUrl = /**
     * @param {?} oAuthPath
     * @param {?} callbackUrl
     * @param {?} windowType
     * @return {?}
     */
    function (oAuthPath, callbackUrl, windowType) {
        var /** @type {?} */ url;
        url = this.atOptions.oAuthBase + "/" + oAuthPath;
        url += "?omniauth_window_type=" + windowType;
        url += "&auth_origin_url=" + encodeURIComponent(callbackUrl);
        if (this.atCurrentUserType != null)
            url += "&resource_class=" + this.atCurrentUserType.name;
        return url;
    };
    /**
     *
     * OAuth
     *
     * @param {?} authWindow
     * @return {?}
     */
    Angular2TokenService.prototype.requestCredentialsViaPostMessage = /**
     *
     * OAuth
     *
     * @param {?} authWindow
     * @return {?}
     */
    function (authWindow) {
        var /** @type {?} */ pollerObserv = interval(500);
        var /** @type {?} */ responseObserv = fromEvent(window, 'message').pipe(pluck('data'), filter(this.oAuthWindowResponseFilter));
        var /** @type {?} */ responseSubscription = responseObserv.subscribe(this.getAuthDataFromPostMessage.bind(this));
        var /** @type {?} */ pollerSubscription = pollerObserv.subscribe(function () {
            if (authWindow.closed)
                pollerSubscription.unsubscribe();
            else
                authWindow.postMessage('requestCredentials', '*');
        });
        return responseObserv;
    };
    /**
     * @param {?} data
     * @return {?}
     */
    Angular2TokenService.prototype.oAuthWindowResponseFilter = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data.message == 'deliverCredentials' || data.message == 'authFailure')
            return data;
    };
    /**
     *
     * Utilities
     *
     * @param {?} name
     * @return {?}
     */
    Angular2TokenService.prototype.getUserTypeByName = /**
     *
     * Utilities
     *
     * @param {?} name
     * @return {?}
     */
    function (name) {
        if (name == null || this.atOptions.userTypes == null)
            return null;
        return this.atOptions.userTypes.find(function (userType) { return userType.name === name; });
    };
    Angular2TokenService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    Angular2TokenService.ctorParameters = function () { return [
        { type: Http },
        { type: ActivatedRoute, decorators: [{ type: Optional }] },
        { type: Router, decorators: [{ type: Optional }] }
    ]; };
    return Angular2TokenService;
}());
export { Angular2TokenService };
function Angular2TokenService_tsickle_Closure_declarations() {
    /** @type {?} */
    Angular2TokenService.prototype.atOptions;
    /** @type {?} */
    Angular2TokenService.prototype.atCurrentUserType;
    /** @type {?} */
    Angular2TokenService.prototype.atCurrentAuthData;
    /** @type {?} */
    Angular2TokenService.prototype.atCurrentUserData;
    /** @type {?} */
    Angular2TokenService.prototype.http;
    /** @type {?} */
    Angular2TokenService.prototype.activatedRoute;
    /** @type {?} */
    Angular2TokenService.prototype.router;
}
