// angular2-token/esm5/a2t-ui/a2t-shared/a2t-form/a2t-form.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable, EventEmitter } from '@angular/core';
import { FormGroup, FormControl } from '@angular/forms';
var A2tFormService = /** @class */ (function () {
    function A2tFormService() {
        this.submit$ = new EventEmitter();
        this.submitLock = false;
    }
    /**
     * @param {?} fields
     * @return {?}
     */
    A2tFormService.prototype.initForm = /**
     * @param {?} fields
     * @return {?}
     */
    function (fields) {
        this.fields = fields;
        this._createFormGroup();
    };
    /**
     * @return {?}
     */
    A2tFormService.prototype.submit = /**
     * @return {?}
     */
    function () {
        this.submitLock = true;
        this.submit$.emit(this.formGroup.value);
    };
    /**
     * @return {?}
     */
    A2tFormService.prototype.unlockSubmit = /**
     * @return {?}
     */
    function () {
        this.formGroup.reset();
        this.submitLock = false;
    };
    ;
    /**
     * @return {?}
     */
    A2tFormService.prototype._createFormGroup = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ group = {};
        this.fields.forEach(function (question) {
            group[question.key] = new FormControl(null, question.validators);
        });
        this.formGroup = new FormGroup(group);
    };
    A2tFormService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    A2tFormService.ctorParameters = function () { return []; };
    return A2tFormService;
}());
export { A2tFormService };
function A2tFormService_tsickle_Closure_declarations() {
    /** @type {?} */
    A2tFormService.prototype.submit$;
    /** @type {?} */
    A2tFormService.prototype.submitLock;
    /** @type {?} */
    A2tFormService.prototype.formGroup;
    /** @type {?} */
    A2tFormService.prototype.fields;
}
