// angular-token/esm2015/lib/angular-token.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function SignInData() { }
if (false) {
    /** @type {?} */
    SignInData.prototype.login;
    /** @type {?} */
    SignInData.prototype.password;
    /** @type {?|undefined} */
    SignInData.prototype.userType;
}
/**
 * @record
 */
export function RegisterData() { }
if (false) {
    /** @type {?} */
    RegisterData.prototype.login;
    /** @type {?} */
    RegisterData.prototype.password;
    /** @type {?} */
    RegisterData.prototype.passwordConfirmation;
    /** @type {?|undefined} */
    RegisterData.prototype.name;
    /** @type {?|undefined} */
    RegisterData.prototype.userType;
}
/**
 * @record
 */
export function RegisterData() { }
/**
 * @record
 */
export function UpdatePasswordData() { }
if (false) {
    /** @type {?} */
    UpdatePasswordData.prototype.password;
    /** @type {?} */
    UpdatePasswordData.prototype.passwordConfirmation;
    /** @type {?|undefined} */
    UpdatePasswordData.prototype.passwordCurrent;
    /** @type {?|undefined} */
    UpdatePasswordData.prototype.userType;
    /** @type {?|undefined} */
    UpdatePasswordData.prototype.resetPasswordToken;
}
/**
 * @record
 */
export function ResetPasswordData() { }
if (false) {
    /** @type {?} */
    ResetPasswordData.prototype.login;
    /** @type {?|undefined} */
    ResetPasswordData.prototype.userType;
}
/**
 * @record
 */
export function AuthData() { }
if (false) {
    /** @type {?} */
    AuthData.prototype.accessToken;
    /** @type {?} */
    AuthData.prototype.client;
    /** @type {?} */
    AuthData.prototype.expiry;
    /** @type {?} */
    AuthData.prototype.tokenType;
    /** @type {?} */
    AuthData.prototype.uid;
}
/**
 * @record
 */
export function UserData() { }
if (false) {
    /** @type {?} */
    UserData.prototype.id;
    /** @type {?} */
    UserData.prototype.provider;
    /** @type {?} */
    UserData.prototype.uid;
    /** @type {?} */
    UserData.prototype.name;
    /** @type {?} */
    UserData.prototype.nickname;
    /** @type {?} */
    UserData.prototype.image;
    /** @type {?} */
    UserData.prototype.login;
}
/**
 * @record
 */
export function UserType() { }
if (false) {
    /** @type {?} */
    UserType.prototype.name;
    /** @type {?} */
    UserType.prototype.path;
}
/**
 * @record
 */
export function AngularTokenOptions() { }
if (false) {
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.angularTokenOptionsProvider;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.apiBase;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.apiPath;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.signInPath;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.signInRedirect;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.signInStoredUrlStorageKey;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.signOutPath;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.validateTokenPath;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.signOutFailedValidate;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.deleteAccountPath;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.registerAccountPath;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.registerAccountCallback;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.updatePasswordPath;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.resetPasswordPath;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.resetPasswordCallback;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.userTypes;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.loginField;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.oAuthBase;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.oAuthPaths;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.oAuthCallbackPath;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.oAuthWindowType;
    /** @type {?|undefined} */
    AngularTokenOptions.prototype.oAuthWindowOptions;
}
