// angular2-token/esm2015/a2t-ui/a2t-shared/a2t-links/a2t-links.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input } from '@angular/core';
export class A2tLinksComponent {
    constructor() { }
}
A2tLinksComponent.decorators = [
    { type: Component, args: [{
                selector: 'a2t-links',
                template: `
        <div class="a2t-wrapper">
            <p><a routerLink="/session/reset-password" *ngIf="case != 'reset-password'">Forgot Password?</a></p>
            <p><a routerLink="/session/sign-up" *ngIf="case != 'sign-up'">Sign Up</a></p>
            <p><a routerLink="/session/sign-in" *ngIf="case != 'sign-in'">Sign In</a></p>
        </div>
    `,
                styles: [`
        .a2t-wrapper {
            margin-top: 20px;
        }

        p {
            margin-bottom: 0;
        }

        a {
            color: #eee !important;
            transition: .3s;
            text-decoration: none;
            font-size: 15px;
            font-weight: 300;
            font-family: "Segoe UI", "Helvetica Neue", Arial, sans-serif;
        }

        a:hover {
            color: white;
        }
    `]
            },] },
];
/** @nocollapse */
A2tLinksComponent.ctorParameters = () => [];
A2tLinksComponent.propDecorators = {
    case: [{ type: Input }]
};
function A2tLinksComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    A2tLinksComponent.prototype.case;
}
