// angular2-token/esm5/a2t-ui/a2t-shared/a2t-form/a2t-form-field/a2t-form-field.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { FormGroup } from '@angular/forms';
import { BaseField } from '../../../a2t-ui.forms';
var A2tFormFieldComponent = /** @class */ (function () {
    function A2tFormFieldComponent() {
    }
    /**
     * @return {?}
     */
    A2tFormFieldComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this._control = this.form.controls[this.question.key];
    };
    Object.defineProperty(A2tFormFieldComponent.prototype, "isValid", {
        get: /**
         * @return {?}
         */
        function () {
            return this._control.valid;
        },
        enumerable: true,
        configurable: true
    });
    A2tFormFieldComponent.decorators = [
        { type: Component, args: [{
                    selector: 'a2t-form-field',
                    template: "\n        <div class=\"a2t-input-group\"\n            [formGroup]=\"form\">\n\n            <label\n                [attr.for]=\"question.key\"\n                *ngIf=\"_control.pristine\">\n                {{question.label}}\n            </label>\n\n            <label class=\"a2t-error\"\n                [attr.for]=\"question.key\"\n                *ngIf=\"_control.hasError('required') && !_control.pristine\">\n                {{question.label}} is required\n            </label>\n\n            <label class=\"a2t-error\"\n                [attr.for]=\"question.key\"\n                *ngIf=\"_control.hasError('minlength')\">\n                {{question.label}} is too short\n            </label>\n\n            <label class=\"a2t-error\"\n                [attr.for]=\"question.key\"\n                *ngIf=\"_control.hasError('maxlength')\">\n                {{question.label}} is too long\n            </label>\n\n            <label class=\"a2t-valid\"\n                [attr.for]=\"question.key\"\n                *ngIf=\"_control.valid && !_control.pristine\">\n                {{question.label}}\n            </label>\n\n            <input\n                [formControlName]=\"question.key\"\n                [id]=\"question.key\"\n                [type]=\"question.type\">\n        </div>\n    ",
                    styles: ["\n        .a2t-input-group {\n            padding-bottom: 40px;\n            padding-right: 20px;\n            padding-left: 20px;\n            font-family: \"Segoe UI\", \"Helvetica Neue\", Arial, sans-serif;\n        }\n\n        .a2t-input-group input {\n            width: 100%;\n            outline: none;\n            border: none;\n            background-color: #eee;\n            line-height: 40px;\n\n            padding-left: 10px;\n            padding-right: 10px;\n        }\n\n        .a2t-input-group label {\n            color: #666;\n            font-weight: 600;\n            font-size: 13px;\n            margin-bottom: 0;\n        }\n\n        .a2t-error {\n            color: #df6564 !important;\n        }\n\n        .a2t-valid {\n            color: #72c380 !important;\n        }\n    "]
                },] },
    ];
    A2tFormFieldComponent.propDecorators = {
        question: [{ type: Input }],
        form: [{ type: Input }]
    };
    return A2tFormFieldComponent;
}());
export { A2tFormFieldComponent };
function A2tFormFieldComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    A2tFormFieldComponent.prototype.question;
    /** @type {?} */
    A2tFormFieldComponent.prototype.form;
    /** @type {?} */
    A2tFormFieldComponent.prototype._control;
}
