// angular2-token/esm2015/a2t-ui/a2t-shared/a2t-form/a2t-form.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable, EventEmitter } from '@angular/core';
import { FormGroup, FormControl } from '@angular/forms';
export class A2tFormService {
    constructor() {
        this.submit$ = new EventEmitter();
        this.submitLock = false;
    }
    /**
     * @param {?} fields
     * @return {?}
     */
    initForm(fields) {
        this.fields = fields;
        this._createFormGroup();
    }
    /**
     * @return {?}
     */
    submit() {
        this.submitLock = true;
        this.submit$.emit(this.formGroup.value);
    }
    /**
     * @return {?}
     */
    unlockSubmit() {
        this.formGroup.reset();
        this.submitLock = false;
    }
    ;
    /**
     * @return {?}
     */
    _createFormGroup() {
        let /** @type {?} */ group = {};
        this.fields.forEach(question => {
            group[question.key] = new FormControl(null, question.validators);
        });
        this.formGroup = new FormGroup(group);
    }
}
A2tFormService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
A2tFormService.ctorParameters = () => [];
function A2tFormService_tsickle_Closure_declarations() {
    /** @type {?} */
    A2tFormService.prototype.submit$;
    /** @type {?} */
    A2tFormService.prototype.submitLock;
    /** @type {?} */
    A2tFormService.prototype.formGroup;
    /** @type {?} */
    A2tFormService.prototype.fields;
}
